Adding QA jobs and expanding test jobs for CI/CD.

tegefaulkes opened the pull request:
### Description

This PR updates the gitlab CI/CD with new jobs for testing if the nix-build outputs run.

Note the pinning of `node-gyp-build` dependency discussed here: https://github.com/MatrixAI/TypeScript-Demo-Lib/pull/35#issuecomment-973852990.

Native dependencies are flaky and has to be considered specially when upgrading.

### Issues Fixed

* Fixes #231
* Fixes #300
* Fixes https://github.com/MatrixAI/Polykey/issues/4
* Fixes #293 
* Fixes #264 
* Fixes #250
* Fixes #313

### Tasks

* [x] Integrate typescript cached compiler path to the `.gitlab-ci.yml` is that ts-node cache is reused. https://github.com/MatrixAI/js-polykey/pull/296#issuecomment-991861764
* [x] Replace `testBinUtils.pkAgent` with `testUtils.setupGlobalAgent`
* [x] See if `testUtils.setupGlobalKeyPair` is sufficient, in which case no need to share a global agent, it will be faster
  - Turns out that by using the global keypair, it takes 3 seconds to create a PK agent, this is sufficient for many tests and they don't need to synchronous on the global PK agent
* ~[ ] Add back in worker manager related testing~ - KeyManager already has tests for worker manager, CLI cannot test worker manager due to weird threadsjs behaviour when spawned as child process inside jest, and the last would just be DB functions which is already tested in js-db
  - [x] Integrate WorkerManager back into `pk agent start`, by using a `--cores` parameter
  - ~[ ] Test the `--workers` parameter, by default it's set to all cores, need to ensure that we are using just 1 worker for testing~ - cannot use workers in `pkSpawn` thus bin tests are not able to test it at all
* [x] Add timeout testing to networking domain
* [x] Find out why `allowHalfOpen` isn't working in utp-native - https://github.com/mafintosh/utp-native/issues/41#issuecomment-993202096
* [x] Fix partial start and partial create for `PolykeyAgent`, `vaults` and `nodes`
* [x] Need to figure out how best to split up tests, assigning arbitrary number of jobs to all the different domains seems brittle.
* [x] Integrate NPM, jest and ts-node caching into Gitlab CI/CD
* [x] Furthermore need to clean up the `.gitlab-ci.yml` file not to have the special case for this branch as it will be merged into master.
* [x] Need to test the actual build to see this working, as well as commentary on the `node-gyp-build` dependency
  - It is expected that nodes may be broken, but this shouldn't affect the build

### Final checklist
* [x] Lint fixed
* [x] Squash and rebased
* [x] Sanity check the final build

tegefaulkes committed:
Adding quality stage to sanity check release builds and parallelising test jobs for CI/CD

CMCDragonkai commented on scripts/runDocker.sh:
```diff
@@ -0,0 +1,4 @@
+#!/bin/sh
+
+NAME=`docker load -i builds/*docker* | awk '{ print $3}'`
```
Need to investigate how this can be done inline.

tegefaulkes replied:
This has been fixed and the old script removed.

CMCDragonkai commented on package.json:
```diff
@@ -52,13 +52,15 @@
   "scripts": {
     "build": "tsc -p ./tsconfig.build.json",
-    "postbuild": "cp -fR src/proto dist",
+    "postbuild": "cp -fR src/proto dist; cp src/notifications/*.json dist/notifications/; cp src/claims/*.json dist/claims/",
```
This needs to be in our developer documentation.

CMCDragonkai commented on src/agent/GRPCClientAgent.ts:
```diff
@@ -72,20 +74,32 @@ class GRPCClientAgent extends GRPCClient<AgentServiceClient> {
 
   @ready(new grpcErrors.ErrorGRPCClientNotStarted())
-  public vaultsGitInfoGet(...args) {
+  public vaultsGitInfoGet(
```
I think there's a robust solution possibly here: https://github.com/MatrixAI/js-polykey/issues/200. Should investigate that here.

CMCDragonkai replied:
Note that the https://gitlab.com/MatrixAI/Engineering/Polykey/js-polykey/-/merge_requests/213 has moved the grpc types into `grpc/types.ts` and it also has `*Client` versions of those stream objects.

tegefaulkes replied:
Noted.

CMCDragonkai commented on src/bin/agent/start.ts:
```diff
@@ -15,6 +15,7 @@ const start = binUtils.createCommand('start', {
   passwordFile: true,
 });
 start.option('-b, --background', 'Starts the agent as a background process');
+start.option('-t --test', 'starts and stops the agent as a quick sanity test.');
```
Should replace this with a background and then in the gitlab YAML directly do a `pk agent stop`. You may need to "sleep" for 1 second.

tegefaulkes replied:
I think it's better to have the test option in agent start but I can remove the flag and have the option hidden as an ENV variable. the test option would also be a very useful debug command down the line.

starting the agent in the background and stopping it complicates the test script a fair bit and I think would be unreliable.

CMCDragonkai replied:
Ok, a sleep might be flaky.

Let's think about the agent testing holistically later. For now a simple sanity check is fine. So running the help page is good enough.

But adding a `--test` flag is not a good idea. It doesn't fit in the rest of the API.

tegefaulkes replied:
I can change it so we can only set it with an env to make it semi-hidden.

tegefaulkes replied:
It's just running the help page now.

CMCDragonkai commented on .gitlab-ci.yml:
```diff
@@ -39,18 +206,120 @@ nix-dry:
+
+running macos build:
+  tags: [ shared-macos-amd64 ]
+  image: macos-11-xcode-12
```
Note that this is a VM image, not a container image. See all possible images here: https://docs.gitlab.com/ee/ci/runners/runner_cloud/linux_runner_cloud.html

tegefaulkes replied:
Noted.

CMCDragonkai commented on .gitlab-ci.yml:
```diff
@@ -39,18 +206,120 @@ nix-dry:
+    - "./builds/*win32* agent start -np ./nodePath --password-file ./testPassword --test"
+
+running macos build:
+  tags: [ shared-macos-amd64 ]
```
Make this consistent with the other `tags` specifiers.

tegefaulkes replied:
Fixed.

CMCDragonkai commented on .gitlab-ci.yml:
```diff
@@ -39,18 +206,120 @@ nix-dry:
+  only:
+    - master
+  tags:
+    - linux
```
Should investigate whether specifying a single container image like Alpine linux is better than just saying any linux image.

tegefaulkes replied:
I'm looking into this but it's hard to find useful information relating to this.
We can specify the image so we have a linux runner but would this be a VM or a docker image? how do we specify a VM image? how do we find and specify a windows image?

so far as I can find the tags are the recommended method but this information for this use case is light. so far the tags have worked fine and I think should continue to work for or purposes.

CMCDragonkai replied:
Linux would always use a container image. In fact the usage of `image` option at the root of the `.gitlab-ci.yml` file means that the container image is used by default across all jobs.

CMCDragonkai replied:
See: https://docs.gitlab.com/ee/ci/quick_start/#gitlab-ciyml-tips and https://docs.gitlab.com/ee/ci/yaml/index.html

CMCDragonkai replied:
I believe you shouldn't even need to specify `tags: linux` since it should be the default runner being used. Plus https://docs.gitlab.com/ee/ci/runners/runner_cloud/linux_runner_cloud.html this indicates that `image` is always used by the linux runners, which means it always means a container image.

This basically means that our `.gitlab-ci.yml` file has to change the `image` setting. For some of the jobs, we can put `image: registry.gitlab.com/matrixai/engineering/maintenance/gitlab-runner` which will use our runner. This should be used for `lint`, `test` and `nix-dry`, `nix`, `packages` as usual. But your new jobs will need to use `image: alpine:3.15` or `alpine:latest`.

Then for your windows and mac runners, the `image` probably means VM and not container.

CMCDragonkai replied:
For windows: https://docs.gitlab.com/ee/ci/runners/runner_cloud/windows_runner_cloud.html

> The Windows runner virtual machine instances do not use the GitLab Docker executor. This means that you can’t specify image or services in your pipeline configuration.

Available software already installed on windows are documented here: https://gitlab.com/gitlab-org/ci-cd/shared-runners/images/gcp/windows-containers/blob/main/cookbooks/preinstalled-software/README.md

This means Windows runners are natively executing the code, no VM and no containers are used.

In that list it shows that chocolatey exists, that would allow us to install other windows software. But think of it as a base windows install, and therefore our single executable should just work there.

CMCDragonkai replied:
And for maos, the `image` is specified. But it means the VM instead of container: https://docs.gitlab.com/ee/ci/runners/runner_cloud/macos/environment.html#vm-images

tegefaulkes replied:
This has been resolved.

CMCDragonkai commented on .gitlab-ci.yml:
```diff
@@ -39,18 +206,120 @@ nix-dry:
+  variables:
+    DOCKER_TLS_CERTDIR: "/certs"
+  services:
+    - docker:19.03.12-dind
```
Should review this in relation to https://docs.gitlab.com/ee/ci/docker/using_docker_build.html#use-the-docker-executor-with-the-docker-image-docker-in-docker

tegefaulkes replied:
In https://docs.gitlab.com/ee/ci/docker/using_docker_build.html#docker-in-docker-with-tls-enabled step 3 is the example I used for this.

CMCDragonkai committed:
Allow STARTING or LIVE for potentially slow command in `tests/bin/agent/status.test.ts`

CMCDragonkai committed:
Make `tests/keys/utils.test.ts` use 1024 size keys

CMCDragonkai committed:
KeyManager tests can use 1024 size keys, and removed asynchronous bug in change password test

CMCDragonkai committed:
The `tests/agent` now uses the global shared key

CMCDragonkai committed:
Fixing tests, acl, agent, bin, bootstrap tests passing

CMCDragonkai commented on src/network/ConnectionForward.ts:
```diff
@@ -239,8 +239,9 @@ class ConnectionForward extends Connection {
 
   protected startTimeout() {
     this.timeout = setTimeout(() => {
+      this.logger.warn('ConnectionForward timing out');
```
What does this mesage convey?

CMCDragonkai replied:
Removed this for now, as this should be done in #304 (error chaining) and https://github.com/MatrixAI/js-logger/issues/3 (structured logging). The combination of the 2 should make these errors more obvious. I've also embedded the commentary for each error into the description of each exception class now in `network/errors` so that way we get useful feedback.

CMCDragonkai commented on src/network/ConnectionForward.ts:
```diff
@@ -239,8 +239,9 @@ class ConnectionForward extends Connection {
+      this.logger.warn('ConnectionForward timing out');
       this.tlsSocket.emit('error', new networkErrors.ErrorConnectionTimeout());
-    }, this.timeoutTime);
+    }, 1000);
```
We should probably have separate branches if we are going to be doing some temporary messing around.

joshuakarp replied:
This is my bad. Feel free to revert my WIP commit on this branch. They're the only temp changes I made.

CMCDragonkai committed:
Splitting the client and agent service RPC handlers

CMCDragonkai committed:
Fixing tests, grpc passing

CMCDragonkai committed:
Adapting seed nodes test to global agent

CMCDragonkai committed:
Fixing tests, notifications, sessions and sigchain is working

joshuakarp committed:
Adding comment about empty default seed nodes in pkStdio, pkExec, etc, and defaulting as empty in setupGlobalAgent

joshuakarp committed:
Splitting nodes.test.ts

CMCDragonkai committed:
Refactoring ForwardProxy and ReverseProxy with more expansive error handling and timeout handling

CMCDragonkai committed:
Integrating new js-async-init with locking start & stop & destroy, blocking ready decorators, and adjustments to partial starts
This requires swapping to explicit partial close due to js-async-init locks
Network proxies is updated to make use of the new js-async-init to address race conditions

CMCDragonkai committed:
Make getLatestClaimId protected method as it is used before Sigchain is ready

CMCDragonkai committed:
Sigchain.addClaim now returns ClaimId as well

CMCDragonkai committed:
Integrating WorkerManager into `pk agent start` with `--workers`
The `--workers` option does not work in tests atm
Also changed seed nodes keyword to `<defaults>`
Fixed tests according to new @matrixai/async-init integration
Removed `setupRemoteKeynode` in favour of `PolykeyAgent`
Replaced `addRemoteDetails` with `nodesConnect` in `testsNodesUtils`
Tests involving status is not robust due to polling and lack of file locking
Removed useless dummy mainnet and testnet nodes

CMCDragonkai committed:
Added section on packaging cross-platform executables

CMCDragonkai committed:
Lintfixed

CMCDragonkai committed:
Integrate child dynamic pipelines for parallelising tests in CI/CD

CMCDragonkai committed:
Integrating ts-node cache, npm cache, and using `npm ci` instead of `npm install`

CMCDragonkai committed:
Integrating fd-lock to Status for serialising readStatus and writeStatus as well as file watching instead of polling for `waitFor`

CMCDragonkai committed:
Split tests according to test directories recursively

CMCDragonkai committed:
Set jest cache to `tmp/jest` and integrate it into CI/CD

CMCDragonkai committed:
Fixing ts-node-cache in child pipeline by not inheriting variables from the parent pipeline

CMCDragonkai committed:
Status on remote live agent test needs node path during CI/CD testing

CMCDragonkai committed:
Removed qa-testing scaffolding for .gitlab-ci.yml

CMCDragonkai committed:
Use --public on pkg to ensure that pkg accepts GPL-3.0 source code
